apps/builder/app/routes/oauth.ws.token.ts

```diff
@@ -0,0 +1,156 @@
+import { type ActionFunctionArgs, json } from "@remix-run/server-runtime";
+import { z } from "zod";
+
+import createDebug from "debug";
+import { fromError } from "zod-validation-error";
+import env from "~/env/env.server";
+import {
+  createAccessToken,
+  readAccessToken,
+  readCodeToken,
+  verifyChallenge,
+} from "~/services/token.server";
+import { isUserAuthorizedForProject } from "~/services/builder-access.server";
+
+/**
+ * OAuth 2.0 Token Request
+ *
+ * https://datatracker.ietf.org/doc/html/rfc7636#section-4.5
+ *
+ * https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.3
+ */
+const TokenRequestSchema = z.object({
+  // Check that the grant_type parameter is present and is one of the supported values
+  grant_type: z.literal("authorization_code"),
+  code: z.string(),
+  redirect_uri: z.string().url(),
+  code_verifier: z.string(),
+  client_id: z.string(),
+  client_secret: z.string(),
+});
+
+const debug = createDebug("OAuth").extend("Token");
+
+/**
+ * OAuth 2.0 Token Request
+ *
+ * https://datatracker.ietf.org/doc/html/rfc7636
+ *
+ * https://datatracker.ietf.org/doc/html/rfc6749
+ */
+export const action = async ({ request }: ActionFunctionArgs) => {
+  debug("Token request received");
+  const jsonBody = Object.fromEntries((await request.formData()).entries());
+  debug("Token request received", jsonBody);
```

Comment:
why do you need 2 debug statements

Reply:
formData can fail

Reply:
is it happening often? I thought it would be only the case when data is entirely broken ... 
if its an absolute edge case I would still not add 2 debug statements permanentlt